### PR TITLE
feat(propagation-utils): end pub-sub process span on promise settled

### PR DESCRIPTION
**Pull request by nordfjord**
When promises are returned from a mapping function we now
end the span when that promises resolves rather than doing
so immediately.

This makes sure that the processing span wraps the entirety of the
processing

## Which problem is this PR solving?

- The pubsub propagation module would end spans prematurely when promises were involved.

An example of what this looks like:
Screen Shot 2022-06-11 at 2 17 00 PM

## Short description of the changes

- Handle promises from mapping functions

## Checklist

- [ ] Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable.

### CODE REVIEW

**Comment by blumamir** [comment]
@nordfjord thank you for the contribution :)
can you please run `npm run lint:fix` to revert all the linter changes? It will really help to focus on the logical changes while reviewing and make CI happy.

Also please add relevant unit tests for this change. Either in the `propagation-utils` package or in the `aws-sdk` instrumentation package which is consuming these utils. Thank you!

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (blumamir) [inline packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts:25]: `Attributes` was added to `@opentelemetry/api` in version `1.1.0` [here](https://github.com/open-telemetry/opentelemetry-js-api/releases/tag/v1.1.0).
The current dependency is on `^1.0.0` which is not compatible. We should either update the dependency or revert to use `SpanAttributes`

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (blumamir) [inline packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts:60]: We prefer to use `===` instead of `==`
optional nit: I think it can be shortened to `typeof (value as any)?.then === 'function'`.

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (blumamir) [inline packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts:59]: Shouldn't the return type be `boolean` instead of `value is Promise<unknown>`

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (blumamir) [inline packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts:63]: Our convention for function names is camelCase: `promiseFinally`.
Also, could be nice to align the style with the rest of the functions in this package, e.g: 
```js
const promiseFinally<T> = (p: Promise<T>, onFinally: () => any): Promise<T> => {...}
```

Comment (blumamir) [reply packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts:63]: Is this function being called anywhere?

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (blumamir) [inline packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts:106]: Here we are creating a new promise and returning it instead of the original one.
Since we are only interested in capturing the promise `resolve` / `rejected` event and do not care about modifying the result itself, I think this can be simplified to just hang another callback on the result promise:
```js
    result.then( () => message[END_SPAN_FUNCTION]?.(), () => message[END_SPAN_FUNCTION]?.())
}
return result;
```

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (nordfjord) [reply packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts:59]: it depends. the `value is` lets TypeScript know that if this predicate passes it can treat the value as a Promise.

```ts
if (isPromise(value)) {
  // TypeScript will know that inside this if block the value can be considered a promise for typechecking purposes
  value.then(x => ...) 
}
```

If `boolean` is used as a return value TypeScript will not narrow the type definition.

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (nordfjord) [reply packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts]: Good point. I've changed this.

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (nordfjord) [reply packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts]: It's not. I've removed it

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (nordfjord) [reply packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts]: 🤦 , sorry about that. I changed this because my tooling was saying SpanAttributes were deprecated.

**Comment by nordfjord** [comment]
Thanks for taking the time to review this. I think this is ready for another review now.

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (Flarna) [reply packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts]: I think this might result in unhandled rejections now.
Calling `.then()` returns a `Promise` which is not used at all here. The catch handler throws which results in rejecting this lost promise. In the original code this chained promise was returned and the user catch handler will be called (if present).

try the following with node 14 or newer:
```js
function foo() {
  const p = Promise.reject("catch foo")
  p.then(() => {}, (e) => { throw e })
  return p
}

function bar() {
  const p = Promise.reject("catch bar")
  return p.then(() => {}, (e) => { throw e })
}

(async () => {
  try { await foo() } catch (e) {}
  try { await bar() } catch (e) {}
})()
```

File: packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts
Comment (blumamir) [reply packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts:59]: Cool! I learned something new :)

**Review by blumamir — APPROVED** [review]
LGTM.
After tests are green this can be merged AFAIC

Thank you for contributing this fix

File: packages/opentelemetry-propagation-utils/package.json
Comment (nordfjord) [reply packages/opentelemetry-propagation-utils/package.json]: Alrighty, copied some versions over! 👍

**Comment by nordfjord** [comment]
Not quite sure what was going on but I had to remove the `--require @opentelemetry/contrib-test-utils` part from the package.json test script so that the tests would pass.

I've verified that `yarn test` works locally now.

**Review by Rauno56 — APPROVED** [review]

**Comment by Rauno56** [comment]
`catch {}` is not supported in node 8. Will merge this after https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1065.

**Review by vmarchaud — APPROVED** [review]

**Pull request merged**
